### PR TITLE
fix(ampd): stellar tx id format

### PR DESCRIPTION
**Pull request by milapsheth**
## Todos

- [x] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

### CODE REVIEW

File: ampd/src/stellar/http_client.rs [hunk 1]
```diff
@@ -15,12 +17,41 @@ pub enum Error {
...
+    let tx_id: String = String::deserialize(deserializer)?;
+    tx_id
+        .strip_prefix("0x")
+        .map(String::from)
+        .ok_or(report!(D::Error::custom(Error::FailedTxIdDeserialization)))
+}
+
+impl StellarTxId {
+    pub fn to_message_id(&self, event_index: u32) -> String {
+        format!("0x{}-{}", self.0.clone(), event_index)
```
Comment (cjcobb23) [inline ampd/src/stellar/http_client.rs:41]: Is this the main change, we append 0x?

File: ampd/src/stellar/http_client.rs [hunk 2]
```diff
@@ -15,12 +17,41 @@ pub enum Error {
...
+pub struct StellarTxId(#[serde(deserialize_with = "deserialize_stellar_tx_id")] String);
+
+/// Deserializes a Stellar tx id from a 0x-prefixed tx id.
+fn deserialize_stellar_tx_id<'de, D>(deserializer: D) -> Result<String, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let tx_id: String = String::deserialize(deserializer)?;
+    tx_id
+        .strip_prefix("0x")
```
Comment (cjcobb23) [inline ampd/src/stellar/http_client.rs:34]: why do we strip 0x and then reappend it?

File: ampd/src/stellar/http_client.rs [hunk 3]
```diff
@@ -15,12 +17,41 @@ pub enum Error {
     Client,
     #[error("invalid tx hash")]
     TxHash,
+    #[error("failed to deserialize tx id")]
+    FailedTxIdDeserialization,
+}
+
+#[derive(Debug, Clone, Eq, Hash, Deserialize, Serialize, PartialEq)]
+pub struct StellarTxId(#[serde(deserialize_with = "deserialize_stellar_tx_id")] String);
```
Comment (cjcobb23) [inline ampd/src/stellar/http_client.rs:25]: what about the event index?

**Review by cjcobb23 — CHANGES_REQUESTED** [review]
Can you summarize what this PR is about?

File: ampd/src/stellar/http_client.rs (hunk 3 above)
Comment (cjcobb23) [reply ampd/src/stellar/http_client.rs:25]: nvm I see this is just the tx id

File: ampd/src/stellar/http_client.rs (hunk 1 above)
Comment (cjcobb23) [reply ampd/src/stellar/http_client.rs:41]: nvm, I see what this is for. I don't think we should hardcode the format here like this, just use the existing message id format struct

File: ampd/src/stellar/http_client.rs [hunk 4]
```diff
@@ -15,12 +17,41 @@ pub enum Error {
...
+
+/// Deserializes a Stellar tx id from a 0x-prefixed tx id.
+fn deserialize_stellar_tx_id<'de, D>(deserializer: D) -> Result<String, D::Error>
```
Comment (cjcobb23) [inline ampd/src/stellar/http_client.rs:28]: I think I understand the issue. The voting verifier emitted the tx id with a 0x prefix, but stellar doesn't like that. Is that right? If so, I don't think this is the best way to go about this, I find this approach rather confusing. I think a simpler way would be to just strip off the 0x prefix inside of transaction_response and transaction_responses. Or just change those functions to accept a type that strips the prefix off when converting from string. I find this automatic stripping as part of deserialization confusing.

Comment (milapsheth) [reply ampd/src/stellar/http_client.rs:28]: sorry for lack of info. Was discussing with Haiyi directly. Another [PR](https://github.com/axelarnetwork/axelar-amplifier/pull/647) was merged to unblock the testing. But this PR is meant to be an alternate fix, but it was failing to build due to a async fn requirement on the Deserialize error. Let me know if you know how to solve it

```
error[E0277]: `?` couldn't convert the error to `<__E as broadcaster::_::_serde::Deserializer<'de>>::Error`
  --> ampd/src/stellar/http_client.rs:24:44
   |
24 | #[derive(Debug, Clone, Eq, Hash, Deserialize, Serialize, PartialEq)]
   |                                            ^ the trait `From<error_stack::Report<<__E as broadcaster::_::_serde::Deserializer<'_>>::Error>>` is not implemented for `<__E as broadcaster::_::_serde::Deserializer<'de>>::Error`, which is required by `std::result::Result<StellarTxId, <__E as broadcaster::_::_serde::Deserializer<'de>>::Error>: FromResidual<std::result::Result<Infallible, error_stack::Report<<__E as broadcaster::_::_serde::Deserializer<'_>>::Error>>>`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = note: required for `std::result::Result<StellarTxId, <__E as broadcaster::_::_serde::Deserializer<'de>>::Error>` to implement `FromResidual<std::result::Result<Infallible, error_stack::Report<<__E as broadcaster::_::_serde::Deserializer<'_>>::Error>>>`
   = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `?` couldn't convert the error to `<__D as broadcaster::_::_serde::Deserializer<'de>>::Error`
  --> ampd/src/stellar/http_client.rs:24:44
   |
24 | #[derive(Debug, Clone, Eq, Hash, Deserialize, Serialize, PartialEq)]
   |                                            ^ the trait `From<error_stack::Report<<__D as broadcaster::_::_serde::Deserializer<'_>>::Error>>` is not implemented for `<__D as broadcaster::_::_serde::Deserializer<'de>>::Error`, which is required by `std::result::Result<<http_client::_::<impl Deserialize<'de> for StellarTxId>::deserialize::__Visitor<'de> as broadcaster::_::_serde::de::Visitor<'de>>::visit_seq::__DeserializeWith<'_>, <__D as broadcaster::_::_serde::Deserializer<'de>>::Error>: FromResidual<std::result::Result<Infallible, error_stack::Report<<__D as broadcaster::_::_serde::Deserializer<'_>>::Error>>>`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = note: required for `Result<__DeserializeWith<'_>, <__D as Deserializer<'de>>::Error>` to implement `FromResidual<std::result::Result<Infallible, error_stack::Report<<__D as broadcaster::_::_serde::Deserializer<'_>>::Error>>>`
   = note: the full name for the type has been written to '/Users/milap/work/axelar-amplifier/target/debug/deps/ampd-84bf7f1394b45ebf.long-type-9934518599439574020.txt'
   = note: consider using `--verbose` to print the full type name to the console
   = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `<D as broadcaster::_::_serde::Deserializer<'de>>::Error` cannot be sent between threads safely
  --> ampd/src/stellar/http_client.rs:32:58
   |
32 |     let tx_id: String = String::deserialize(deserializer)?;
   |                                                          ^ `<D as broadcaster::_::_serde::Deserializer<'de>>::Error` cannot be sent between threads safely
   |
   = help: the trait `std::marker::Send` is not implemented for `<D as broadcaster::_::_serde::Deserializer<'de>>::Error`, which is required by `std::result::Result<std::string::String, error_stack::Report<<D as broadcaster::_::_serde::Deserializer<'de>>::Error>>: FromResidual<std::result::Result<Infallible, <D as broadcaster::_::_serde::Deserializer<'_>>::Error>>`
   = note: required for `<D as broadcaster::_::_serde::Deserializer<'de>>::Error` to implement `error_stack::Context`
   = note: required for `error_stack::Report<<D as broadcaster::_::_serde::Deserializer<'de>>::Error>` to implement `From<<D as broadcaster::_::_serde::Deserializer<'_>>::Error>`
   = note: required for `std::result::Result<std::string::String, error_stack::Report<<D as broadcaster::_::_serde::Deserializer<'de>>::Error>>` to implement `FromResidual<std::result::Result<Infallible, <D as broadcaster::_::_serde::Deserializer<'_>>::Error>>`
help: consider further restricting the associated type
   |
30 |     D: Deserializer<'de>, <D as broadcaster::_::_serde::Deserializer<'de>>::Error: std::marker::Send
   |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0277]: `<D as broadcaster::_::_serde::Deserializer<'de>>::Error` cannot be shared between threads safely
```

Comment (milapsheth) [reply ampd/src/stellar/http_client.rs:28]: And to your larger point about the design, I think we need to fix the poll started event to not split tx id and event index. We should keep it as message id, and the appropriate handler should convert to the MessageIdFormat itself

**Review by haiyizxx — APPROVED** [review]

**Comment by milapsheth** [comment]
this was addressed by emitting `message_id` in the poll events
